scripts/elasticsearch_nodes.py

```diff
@@ -0,0 +1,127 @@
+max_attempts = 10
+
+
+def main() -> None:
+    node_ids = get_node_ids()
+    data = prepare_node_ids(node_ids)
+    print(json.dumps(data))
+
+
+def get_node_ids() -> Sequence[str]:
+    """
+    Return a list of ES node IDs used by the current deployment.
+    """
+    for i in range(max_attempts):
+        log.debug(f'Attempt {i}/{max_attempts}')
+        try:
+            response = aws.cloudwatch.list_metrics(Namespace='AWS/ES',
+                                                   MetricName='CPUUtilization',
+                                                   RecentlyActive='PT3H',
+                                                   Dimensions=[
+                                                       {
+                                                           'Name': 'DomainName',
+                                                           'Value': config.es_domain
+                                                       }
+                                                   ])
+        except ClientError as e:
+            log.warning(e.response['Error']['Code'] + '. Retrying...')
+        else:
+            nodes = [
+                d['Value']
+                for m in response['Metrics']
+                for d in m['Dimensions']
+                if d['Name'] == 'NodeId'
+            ]
+            if len(nodes) > 0:
+                return nodes
+            else:
+                log.warning('No node IDs found in response.')
+        sleep(5)
+
+    raise RuntimeError('Failed to obtain node IDs. Max attempts exhausted.')
+
+
+def prepare_node_ids(node_ids: Sequence[str]) -> Mapping[str, str]:
+    """
+    Return the given node IDs in a data structure that TerraForm can use as an
+    external data source.
+    """
+    # The list of nodes IDs is appended with dummy entries to ensure the length
+    # matches the number of placeholders in the template. If our list had less
+    # than the template, Terraform would throw an "Invalid index" error. Also,
+    # the dummy entries cannot be empty strings.
+    placeholder_count = node_placeholder_count()
+    require(placeholder_count > 0,
+            'Error: No node ID placeholders found in template file.')
+    node_ids = sorted(node_ids)
+    node_ids += list(repeat('blank', placeholder_count - len(node_ids)))
+
+    # As required by Terraform for an external data source, the output is
+    # formatted as a map of string values.
+    return {
+        "nodes": ','.join(node_ids)
+    }
+
+
+def node_placeholder_count() -> int:
+    """
+    Return the number of unique ES node ID placeholders found in the cloudwatch
+    dashboard template file. (i.e. the largest ${local.nodes[0]} index value +1)
+    """
+    with open(config.cloudwatch_dashboard_template) as f:
+        body = json.load(f)
+
+    jq_filters = [
+        '.widgets[].properties.metrics[]?[-4]',
+        'values',
+        r'capture("local\\.nodes\\[(?<a>[0-9]+)\\]").a',
```

Comment:
I'm confused, the point of r-strings is so you don't have to double the backslashes.

Reply:
Without an r-string, 3 backslashes would be required.

```
>>> import jq
>>> jq.all(r'.[]|capture("x(?<a>\\d+)").a', ['xa', 'x0', '0x', 'x99'])
['0', '99']
>>> jq.all('.[]|capture("x(?<a>\\\d+)").a', ['xa', 'x0', '0x', 'x99'])
['0', '99']
```

Reply:
> On why '\\\d' is the same as '\\\\d' and why it's discouraged

This is very interesting. Thank you for the explanation.